suseconnect/migration.go

```diff
@@ -0,0 +1,566 @@
+package main
+
+// TODO LIST
+// * zypp_backup/zypp_restore functions
+// * zypper dup wrapper with options (these are mostly pass-through from plugin args)
+// *   passthrough zypper dup options:
+// *     --allow-vendor-change
+// *     --from
+// *     --repo
+// *     --debug-solver
+// *     --recommends
+// *     --no-recommends
+// *     --replacefiles
+// *     --details
+// *     --download (including --download-only)
+// * --selfupdate option
+// * --query plugin option
+// * --break-my-system option
+// * --product option (offline migration)
+// * obsolete repo disabling (including --disable-repos option)
+// * interactive migration mode
+// * snapshots (snapper wrapper)
+// * Leap -> SLES migration case
+
+import (
+	_ "embed"
+	"errors"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"os/signal"
+	"sort"
+	"syscall"
+
+	"github.com/SUSE/connect-ng/internal/connect"
+)
+
+var (
+	//go:embed migrationUsage.txt
+	migrationUsageText string
+	// flag indicating interuption by INT/TERM signal
+	interrupted bool
+
+	// ErrInterrupted is returned when execution was interrupted by INT/TERM signal
+	ErrInterrupted = errors.New("Interrupted")
+)
+
+// logger shortcuts
+var (
+	Debug    *log.Logger = connect.Debug
+	QuietOut *log.Logger = connect.QuietOut
+)
+
+func migrationMain() {
+	var (
+		dummy             bool
+		verbose           bool
+		quiet             bool
+		nonInteractive    bool
+		autoAgreeLicenses bool
+		noSnapshots       bool
+		noSelfUpdate      bool
+		migrationNum      int
+		fsRoot            string
+	)
+
+	flag.Usage = func() {
+		fmt.Print(migrationUsageText)
+	}
+	// dummy flag to keep default but accept cli arg
+	flag.BoolVar(&dummy, "no-verbose", false, "")
+	flag.BoolVar(&verbose, "verbose", false, "")
+	flag.BoolVar(&verbose, "v", false, "")
+	// dummy flag to keep default but accept cli arg
+	flag.BoolVar(&dummy, "no-quiet", false, "")
+	flag.BoolVar(&quiet, "quiet", false, "")
+	flag.BoolVar(&quiet, "q", false, "")
+	flag.BoolVar(&nonInteractive, "non-interactive", false, "")
+	flag.BoolVar(&nonInteractive, "n", false, "")
+	flag.BoolVar(&autoAgreeLicenses, "auto-agree-with-licenses", false, "")
+	flag.BoolVar(&autoAgreeLicenses, "l", false, "")
+	flag.BoolVar(&noSnapshots, "no-snapshots", false, "")
+	// dummy flag to keep default but accept cli arg
+	flag.BoolVar(&dummy, "selfupdate", false, "")
+	flag.BoolVar(&noSelfUpdate, "no-selfupdate", false, "")
+	flag.IntVar(&migrationNum, "migration", 0, "")
+	flag.StringVar(&fsRoot, "root", "", "")
+
+	flag.Parse()
+	// this is only to keep the flag parsing logic simple and avoid double
+	// negations/negatives below
+	selfUpdate := !noSelfUpdate
+
+	if verbose {
+		connect.EnableDebug()
```

Comment:
I don't see the original doing this. It adds a lot of detailed output. But I think we do need some way to enable debug logging on the suseconnect library.

Reply:
this was me trying to do a "clever trick" and map `--verbose` to `--debug` which is not directly present in `zypper-migration-plugin`. the other option would be to add something like `VerboseOut` to match the `QuietOut`.

Reply:
This spews pages and pages of detailed stuff, and makes it very hard to see the messages from `--verbose`. Also my scrollback buffer did not go back far enough.

I think we should add a `--debug` option to the zypper migration plugin CLI to enable this level of logging.

Reply:
Should `--debug` imply `--verbose` ?

Reply:
I changed the `--verbose` flag to support multi-use. Single `--verbose` is like before. Double `--verbose --verbose` enables connect's debug output.

Reply:
Unfortunately `flag` doesn't seem to support GNU style args. There's https://github.com/spf13/pflag but I don't think it makes sense to add a dependency just for this corner case. I'll change it to `--debug` if you think it's more readable.

Reply:
done.